flow/BlobCipher.h

```diff
@@ -0,0 +1,244 @@
+/*
+ * BlobCipher.h
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <memory>
+#if (!defined(TLS_DISABLED) && !defined(_WIN32))
+#define ENCRYPTION_ENABLED 1
+#else
+#define ENCRYPTION_ENABLED 0
+#endif
+
+#if ENCRYPTION_ENABLED
+
+#include "flow/Arena.h"
+#include "flow/FastRef.h"
+#include "flow/flow.h"
+#include "flow/xxhash.h"
+
+#include <openssl/aes.h>
+#include <openssl/engine.h>
+#include <openssl/evp.h>
+#include <openssl/hmac.h>
+#include <openssl/sha.h>
+#include <string>
+#include <vector>
+
+#define AES_256_KEY_LENGTH 32
+#define AES_256_TAG_LENGTH 16
+#define AES_256_IV_LENGTH 16
+
+using BlobCipherDomainId = uint64_t;
+using BlobCipherRandomSalt = uint64_t;
+using BlobCipherBaseKeyId = uint64_t;
+using BlobCipherIV = std::array<unsigned char, AES_256_IV_LENGTH>;
+using BlobCipherChecksum = XXH64_hash_t;
+
+typedef enum { BLOB_CIPHER_ENCRYPT_MODE_NONE = 0, BLOB_CIPHER_ENCRYPT_MODE_AES_256_CTR = 1 } BlockCipherEncryptMode;
+
+// BlobCipher Encryption header format
+// The header is persisted as 'plaintext' for encrypted block containing sufficient information for encyrption key
+// regeneration to assit decryption on reads. The total space overhead is 40 bytes.
+
+#pragma pack(push, 1) // exact fit - no padding
+typedef struct BlobCipherEncryptHeader {
+	union {
+		struct {
+			uint8_t
+			    size; // reading first byte is sufficient to determine header length. ALWAYS THE FIRST HEADER ELEMENT.
+			uint8_t headerVersion{};
+			uint8_t encryptMode{};
+			uint8_t _reserved[5]{};
+		} flags;
+		uint64_t _padding{};
+	};
+	BlobCipherDomainId encryptDomainId{};
+	BlobCipherBaseKeyId baseCipherId{};
+	BlobCipherRandomSalt salt{};
+	BlobCipherChecksum checksum{};
+
+	BlobCipherEncryptHeader();
+} BlobCipherEncryptHeader;
+#pragma pack(pop)
+
+// This interface is in-memory representation of CipherKey used for encryption/decryption information. It caches base
+// encyrption key properties as well as apply HMAC_SHA_256 derivation technique to generate a new encryption key.
+
+class BlobCipherKey : public ReferenceCounted<BlobCipherKey>, NonCopyable {
+	// Encryption domain boundary identifier
+	BlobCipherDomainId encryptDomainId;
+	// Base encyrption cipher key properties
```

Comment:
Nit:typo "encyrption"? Not the only one in this file

Reply:
Thanks for pointing. Fixed!